lib/Sema/TypeCheckDeclOverride.cpp

```diff
@@ -937,11 +937,23 @@ static GenericSignature *getNewGenericSignature(ValueDecl *base,
   auto derivedGenericCtx = derived->getAsGenericContext();
   auto &ctx = base->getASTContext();
 
+  if (!baseGenericCtx) {
+    return nullptr;
+  }
+
   auto baseClass =
       baseGenericCtx->getInnermostTypeContext()->getSelfClassDecl();
   auto derivedClass =
       derivedGenericCtx->getInnermostTypeContext()->getSelfClassDecl();
 
+  if (!derivedClass) {
+    return nullptr;
+  }
+
+  if (!derivedClass->getInterfaceType()->is<LValueType>()) {
```

Comment:
This is always going to be true, because the type of a ClassDecl is never an LValueType.

Reply:
@slavapestov Right, so does that mean we should remove the assert from getContextSubstitutions()? Otherwise it’s going to hit the assert there.

Reply:
The assert in getContextSubstitutionMap() is correct. However getInterfaceType() on a ClassDecl should never return an LValueType. Can you print the decl and its type to verify what's going on?

Reply:
I see the problem. When you compute 'subMap' below, you should use `derivedClass->getSuperclass()` and not `derivedClass->getInterfaceType()`.

The assert that is firing is that you're passing a MetatypeType to getContextSubstitutions(), not an LValueType.

Reply:
@slavapestov Ah ok. But then the code causes a crash while doing `for (auto reqt : baseClassSig->getRequirements()) { ... }` (it's calling `getRequirementsBuffer` which causes the crash). I think it makes sense for `baseClassSig` to be null if the class itself doesn't have any generic parameters:

```swift
protocol SR_4206_P0 {}

protocol SR_4206_P1: SR_4206_P0 {}
protocol SR_4206_P2: SR_4206_P0 {}

class SR_4206_C1 {
  func foo<T: SR_4206_P1>(param: T) {} // expected-note {{overridden declaration is here}}
}

class SR_4206_C2: SR_4206_C1 {
  override func foo<T: SR_4206_P2>(param: T) {} // expected-error {{cannot override method 'foo' with a different generic signature}}
}
```

Should we be looping over the method's generic requirements instead in this case?